tensorflow/core/kernels/svd_op_gpu.cu.cc

```diff
@@ -0,0 +1,249 @@
+    "valid.";
+
+typedef Eigen::GpuDevice GPUDevice;
+
+template <class Scalar>
+class SvdOpGpu : public AsyncOpKernel {
+ public:
+  explicit SvdOpGpu(OpKernelConstruction* context) : AsyncOpKernel(context) {
+    OP_REQUIRES_OK(context, context->GetAttr("compute_uv", &compute_uv_));
+    OP_REQUIRES_OK(context, context->GetAttr("full_matrices", &full_matrices_));
+  }
+
+  void ComputeAsync(OpKernelContext* context, DoneCallback done) final {
+    const Tensor& input = context->input(0);
+    const int ndims = input.dims();
+    const int64 m = input.dim_size(ndims - 2);
+    const int64 n = input.dim_size(ndims - 1);
+    const int64 p = std::min(m, n);
+
+    // This is a limitation of cuSolver
+    OP_REQUIRES_ASYNC(
+        context, m >= n,
+        errors::InvalidArgument(
+            "GPU implementation of SVD only supports m>=n, but m=", m, ", n=",
+            n),
+        done);
+
+    // Validate inputs.
+    OP_REQUIRES_ASYNC(
+        context, ndims >= 2,
+        errors::InvalidArgument("Input must have rank >= 2, got ", ndims),
+        done);
+
+    // output tensors.
+    Tensor* outputU = NULL;
+    Tensor* outputS = NULL;
+    Tensor outputVT;
+    Tensor* outputV = NULL;
+
+    // compute  shapes
+    TensorShape shapeRaw = input.shape();
+    shapeRaw.RemoveDim(shapeRaw.dims() - 1);
+    shapeRaw.RemoveDim(shapeRaw.dims() - 1);
+    TensorShape shapeS = shapeRaw;
+    TensorShape shapeU = shapeRaw;
+    TensorShape shapeVT = shapeRaw;
+    TensorShape shapeV = shapeRaw;
+    shapeS.AddDim(p);
+    if (compute_uv_) {
+      if (full_matrices_) {
+        shapeU.AddDim(m);
+        shapeU.AddDim(m);
+        shapeVT.AddDim(n);
+        shapeVT.AddDim(n);
+        shapeV.AddDim(n);
+        shapeV.AddDim(n);
+      } else {
+        shapeU.AddDim(m);
+        shapeU.AddDim(p);
+        shapeVT.AddDim(p);
+        shapeVT.AddDim(n);
+        shapeV.AddDim(n);
+        shapeV.AddDim(p);
+      }
+    } else {
+      shapeU = TensorShape({0});
+      shapeV = TensorShape({0});
+    }
+
+    // allocate output
+    OP_REQUIRES_OK_ASYNC(context, context->allocate_output(0, shapeS, &outputS),
+                         done);
+    OP_REQUIRES_OK_ASYNC(context, context->allocate_output(1, shapeU, &outputU),
+                         done);
+    OP_REQUIRES_OK_ASYNC(context, context->allocate_output(2, shapeV, &outputV),
+                         done);
+    if (compute_uv_) {
+      OP_REQUIRES_OK_ASYNC(
+          context, context->allocate_temp(input.dtype(), shapeVT, &outputVT),
+          done);
+    }
+
+    if (n == 0 || m == 0) {
+      // If X is an empty matrix (0 rows, 0 col), X * X' == X.
+      // Therefore, we return X.
+      done();
+      return;
+    }
+
+    // Copy and reshape input tensor
+    // SVD modifies the input, so I need to copy it.
+    Tensor inputCopy;
+    OP_REQUIRES_OK_ASYNC(
+        context,
+        context->allocate_temp(input.dtype(), input.shape(), &inputCopy), done);
```

Comment:
You could use buffer forwarding here, like so:
```
    // Reuse the input buffer or make a copy for the SVD depending on whether this ops owns the 
    // input buffer exclusively.
    Tensor input_copy;
    OP_REQUIRES_OK_ASYNC(
        context,
        context->forward_input_or_allocate_temp(
            {0}, DataTypeToEnum<Scalar>::value, input.shape(), &input_copy),
        done);

    if (!input.SharesBufferWith(input_copy)) {
      const GPUDevice& d = context->eigen_device<GPUDevice>();
      d.memcpy(input_copy.flat<Scalar>().data(), input.flat<Scalar>().data(),
               input.NumElements() * sizeof(Scalar));
    }
```

Please do not call cudaMemcpy directly as it synchronizes the stream.